manager/allocator/allocator_new.go

```diff
@@ -0,0 +1,929 @@
+				case api.EventDeleteNode:
+					if ev.Node != nil {
+						a.pendingMu.Lock()
+						a.network.DeallocateNode(ev.Node)
+						delete(a.pendingNodes, ev.Node.ID)
+						a.pendingMu.Unlock()
+					}
+				}
+			}
+		}
+	}()
+
+	log.G(ctx).Debug("starting batch processing loop")
+	// allocations every 500 milliseconds
+	batchTimer := time.NewTimer(maxBatchInterval)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-batchTimer.C:
+				a.pendingMu.Lock()
+				total := len(a.pendingNetworks) + len(a.pendingNodes) + len(a.pendingTasks)
+				a.pendingMu.Unlock()
+				// we release the lock before checking the total, which is fine
+				// because the worst case is something gets deleted and the
+				// processPendingAllocations just jumps straight through
+				if total > 0 {
+					// every batch interval, do all allocations and then reset the
+					// timer to ready for the next batch
+					a.processPendingAllocations(ctx)
+				}
+				batchTimer.Reset(maxBatchInterval)
+			}
+		}
+	}()
+
+	log.G(ctx).Info("allocator is started")
+	wg.Wait()
+	log.G(ctx).Info("allocator is finished")
+	defer log.G(ctx).Debug("all defers exited")
+	return nil
+}
+
+// Stop is a method that stops the running allocator, blocking until it has
+// fully stopped
+func (a *NewAllocator) Stop() {
+	a.stopOnce.Do(func() {
+		close(a.stop)
+	})
+	// NOTE(dperny) an un-selected channel read is inescapable and can result
+	// in deadlocks, however, being able to escape this channel read with, say,
+	// a context cancelation, would just lead to resource leaks of an already
+	// deadlocked allocator, so the risk of a deadlock isn't that important
+	<-a.stopped
+}
+
+func (a *NewAllocator) processPendingAllocations(ctx context.Context) {
+	ctx = log.WithField(ctx, "method", "(*NewAllocator).processPendingAllocations")
+
+	// we need to hold the lock the whole time we're allocating.
+	a.pendingMu.Lock()
+	defer a.pendingMu.Unlock()
+
+	log.G(ctx).Debugf(
+		"processing pending allocations. %v networks, %v nodes, %v tasks",
+		len(a.pendingNetworks), len(a.pendingNodes), len(a.pendingTasks),
+	)
+
+	defer func() {
+		log.G(ctx).Debugf(
+			"finished processing pending allocations. retrying: %v networks, %v nodes, %v tasks",
+			len(a.pendingNetworks), len(a.pendingNodes), len(a.pendingTasks),
+		)
+	}()
+
+	// collect metrics on how long each batch of pending allocations takes.
+	// start at timer after we acquire the lock, and then defer the done call.
+	done := metrics.StartTimer(batchProcessingDuration)
+	defer done()
+
+	// overlayAllocated is a variable that tracks whether a new overlay network
+	// has been allocated. If it is true, then regardless of which nodes are
+	// currently pending, all nodes will need to be reallocated with the new
+	// network. we will do this after network allocation, but before task
+	// allocation, to avoid a case where nodes use up all of the available
```

Comment:
nit:  isn't this the case we want to favor, as opposed to avoid?  We want the nodes to use up all the available network addresses as opposed to tasks using up all of them.